files/image_config/corefile_uploader/core_uploader.py

```diff
@@ -0,0 +1,294 @@
+
+def log_warn(m):
+    log_msg(syslog.LOG_WARNING, this_file, m)
+
+def log_debug(m):
+    log_msg(syslog.LOG_DEBUG, this_file, m)
+
+
+def mkdir(p):
+    l = p.split("/")
+    if p.endswith("/"): 
+        l.pop()
+
+    e = l[0:len(l)]
+
+    while not os.path.exists("/".join(e)):
+        e.pop()
+
+    while len(e) != len(l):
+        e.append(l[len(e)])
+        os.mkdir("/".join(e))
+
+def make_new_dir(p):
+    if os.path.isdir(p):
+        rmtree(p)
+    elif os.path.exists(p):
+        os.path.unlink(p)
+    mkdir(p)
+
+def parse_a_json(data, prefix, val):
+    for i in data:
+        if type(data[i]) == dict:
+            parse_a_json(data[i], prefix + (i,), val)
+        else:
+            val[prefix + (i,)] = data[i]
+
+class config:
+    parsed_data = {}
+    cfg_data = {}
+
+    def __init__(self):
+        while not os.path.exists(RC_FILE):
+            # Wait here until service restart
+            log_err("Unable to retrieve Azure storage credentials")
+            time.sleep (HOURS_4)
+
+        with open(RC_FILE, 'r') as f:
+            self.parsed_data = json.load(f)
+            parse_a_json(self.parsed_data, (), self.cfg_data)
+
+    def get_data(self, k):
+        return self.cfg_data[k] if self.cfg_data.has_key(k) else ""
+
+    def get_dict(self):
+        return self.parsed_data
+
+    def get_core_info(self, corepath, devicename):
+        info = {}
+        info["corefname"] = os.path.basename(corepath)
+        info["tstamp"] = str(os.stat(corepath).st_ctime)
+        info["devicename"] = devicename
+
+        lpath = self.get_data(("metadata_files_in_archive", "core_info"))
+        f = open(lpath, "w+")
+        f.write(json.dumps(info))
+        f.close()
+
+        return lpath
+
+
+class Watcher:
+
+    def __init__(self):
+        self.observer = Observer()
+
+    def run(self):
+        event_handler = Handler()
+        self.observer.schedule(event_handler, CORE_FILE_PATH)
+        self.observer.start()
+        try:
+            while True:
+                time.sleep(5)
+        except:
+            self.observer.stop()
+            log_err("Error in watcher")
+
+        self.observer.join()
+
+def set_env(lst):
+    for k in lst:
+        if lst[k]:
+            os.environ[k] = lst[k]
+            log_debug("set env {} = {}".format(k, lst[k]))
+
+class Handler(FileSystemEventHandler):
+
+    @staticmethod
+    def init():
+        global hostname, sonicversion, asicname, acctname, acctkey, sharename
+        global cwd, cfg
+
+        cfg = config()
+
+        set_env(cfg.get_dict()["env"])
+
+        hostname = socket.gethostname()
+        if not hostname:
+            raise Exception("Failed to read hostname from /etc/sonic/sonic_version.yml")
+
+        acctname = cfg.get_data(("azure_sonic_core_storage", "account_name"))
+        acctkey = cfg.get_data(("azure_sonic_core_storage", "account_key"))
+        sharename = cfg.get_data(("azure_sonic_core_storage", "share_name"))
```

Comment:
Would the cfg change underneath if new credentials got pushed down? Should you also watch the config file since you already has a watcher?

Reply:
No, It should not.

The credentials update happens only in two instances:
1) Conversion/Image upgrade
2) Manually controlled through azconfigure.

In either case, I  intentionally expect the caller to restart the service. This is *not* any essential service like PAM auth. So I prefer, an explicitly restart.